src/ripple/nodestore/impl/Shard.cpp

```diff
@@ -40,346 +40,528 @@ Shard::Shard(
-                             << " in function " << __func__;
-        }
+    try
+    {
+        boost::filesystem::remove_all(dir_);
+    }
+    catch (std::exception const& e)
+    {
+        JLOG(j_.fatal()) << "shard " << index_
+                         << " Exception caught in function " << __func__
+                         << ". Error: " << e.what();
     }
 }
 
 bool
-Shard::open(Scheduler& scheduler, nudb::context& ctx)
+Shard::init(Scheduler& scheduler, nudb::context& context)
 {
-    std::lock_guard lock{mutex_};
-    assert(!backend_);
+    Section section{app_.config().section(ConfigSection::shardDatabase())};
+    std::string const type{get<std::string>(section, "type", "nudb")};
+    auto const factory{Manager::instance().find(type)};
+    if (!factory)
+    {
+        JLOG(j_.error()) << "shard " << index_ << " failed to find factory for "
+                         << type;
+        return false;
+    }
+    section.set("path", dir_.string());
 
-    Config const& config{app_.config()};
+    std::lock_guard lock{mutex_};
+    if (backend_)
     {
-        Section section{config.section(ConfigSection::shardDatabase())};
-        std::string const type{get<std::string>(section, "type", "nudb")};
-        auto factory{Manager::instance().find(type)};
-        if (!factory)
-        {
-            JLOG(j_.error()) << "shard " << index_
-                             << " failed to create backend type " << type;
-            return false;
-        }
+        JLOG(j_.error()) << "shard " << index_ << " already initialized";
+        return false;
+    }
+    backend_ = factory->createInstance(
+        NodeObject::keyBytes, section, scheduler, context, j_);
 
-        section.set("path", dir_.string());
-        backend_ = factory->createInstance(
-            NodeObject::keyBytes, section, scheduler, ctx, j_);
+    return open(lock);
+}
+
+bool
+Shard::isOpen() const
+{
+    std::lock_guard lock(mutex_);
+    if (!backend_)
+    {
+        JLOG(j_.error()) << "shard " << index_ << " not initialized";
+        return false;
     }
 
-    using namespace boost::filesystem;
-    auto preexist{false};
-    auto fail = [this, &preexist](std::string const& msg) {
-        pCache_.reset();
-        nCache_.reset();
-        backend_.reset();
-        lgrSQLiteDB_.reset();
-        txSQLiteDB_.reset();
-        acquireInfo_.reset();
+    return backend_->isOpen();
+}
 
-        if (!preexist)
-            remove_all(dir_);
+bool
+Shard::tryClose()
+{
+    // Keep database open if being acquired or finalized
+    if (state_ != final)
+        return false;
 
-        if (!msg.empty())
-        {
-            JLOG(j_.fatal()) << "shard " << index_ << " " << msg;
-        }
+    std::lock_guard lock(mutex_);
+    if (!backend_)
+    {
+        JLOG(j_.error()) << "shard " << index_ << " not initialized";
+        return false;
+    }
+    if (!backend_->isOpen())
         return false;
```

Comment:
The check is for exiting early on a closed backend, which could occur if `tryClose` had already been called.  Because a closed shard has cleared its caches, a `sweep` call would have no effect. Please let me know if I missed something or misunderstood your comment. Thanks!

Reply:
Sounds like the code is doing the right thing as it is.  It's a legit state (no log write necessary) and an early return is the right thing to do.   Thanks for answering my questions @miguelportilla!